drivers/pwm/pwm_max31790.c

```diff
@@ -0,0 +1,677 @@
+	MAX37190_FANXCONFIGURATION_LOCKEDROTORPOLARITY_SET(value_fan_configuration, 0b0);
+	MAX37190_FANXCONFIGURATION_TACH_SET(value_fan_configuration, 0b0);
+	MAX37190_FANXCONFIGURATION_TACHINPUTENABLED_SET(value_fan_configuration, 0b1);
+
+	MAX37190_FANXDYNAMICS_SPEEDRANGE_SET(value_fan_dynamics, value_speed_range);
+	MAX37190_FANXDYNAMICS_PWMRATEOFCHANGE_SET(value_fan_dynamics, value_pwm_rate_of_change);
+	MAX37190_FANXDYNAMICS_ASYMMETRICRATEOFCHANGE_SET(value_fan_dynamics, 0b1);
+
+	switch (channel) {
+	case 0:
+		register_fan_configuration = MAX37190_FAN1CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT1TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT1TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN1DYNAMICS;
+		break;
+	case 1:
+		register_fan_configuration = MAX37190_FAN2CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT2TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT2TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN2DYNAMICS;
+		break;
+	case 2:
+		register_fan_configuration = MAX37190_FAN3CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT3TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT3TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN3DYNAMICS;
+		break;
+	case 3:
+		register_fan_configuration = MAX37190_FAN4CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT4TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT4TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN4DYNAMICS;
+		break;
+	case 4:
+		register_fan_configuration = MAX37190_FAN5CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT5TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT5TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN5DYNAMICS;
+		break;
+	case 5:
+		register_fan_configuration = MAX37190_FAN6CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT6TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT6TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN6DYNAMICS;
+		break;
+	default:
+		LOG_ERR("invalid channel %i", channel);
+		return -EINVAL;
+	}
+
+	uint16_t pwm_target_duty_cycle =
+		pulse_count * MAX31790_PWMTARGETDUTYCYCLE_MAXIMUM / period_count;
+	uint8_t value_pwm_target_duty_cycle_msb = pwm_target_duty_cycle >> 1;
+	uint8_t value_pwm_target_duty_cycle_lsb = (pwm_target_duty_cycle & 0x01) << 7;
+
+	MAX37190_FANXCONFIGURATION_MODE_SET(value_fan_configuration, 0b0);
+
+	result = max31790_write_register(dev, register_pwm_target_duty_cycle_msb,
+					 value_pwm_target_duty_cycle_msb);
+	if (result != 0) {
+		return result;
+	}
+
+	result = max31790_write_register(dev, register_pwm_target_duty_cycle_lsb,
+					 value_pwm_target_duty_cycle_lsb);
+	if (result != 0) {
+		return result;
+	}
```

Comment:
These are always consecutive, how about writing them in a single i2c transaction?

Reply:
According to the datasheet this IC does not support writing multiple registers at once.

Reply:
![x](https://github.com/zephyrproject-rtos/zephyr/assets/891546/ca282735-87c7-4d1a-b703-d4250e93a5ac)
Doesn't it? The Linux driver also writes in words: https://lxr.missinglinkelectronics.com/linux/drivers/hwmon/max31790.c#L323 Are you even guaranteed atomicity if you write a two bytes values in two separate i2c transactions?

Reply:
Thanks for pointing this out, I couldn't (obviously) find it by myself. At least in the I2C part o the spec it wasn't mentioned, hence I assumed it is not possible.
As the driver locks the accesses with mutexes atomicity is not really an issue, but I for sure makes sense to reduce the amount of I2C transaction. Therefore, I'm gonna combine them anyway.

Reply:
done